parlai/core/torch_agent.py

```diff
@@ -690,6 +690,40 @@ def __init__(self, opt, shared=None):
         self.rank_candidates = opt['rank_candidates']
         self.add_person_tokens = opt.get('person_tokens', False)
 
+    @classmethod
+    def upgrade_opt(cls, opt_on_disk):
+        """
+        Upgrade legacy options when loading an opt file from disk.
+
+        When loading a previously trained model file, we may wish to ensure
+        backwards compatibility. For example, we may introduce a new option
+        today that wasn't available initially, but we want the default on all
+        new models to be the new behavior.
+
+        ``upgrade_opt`` provides an opportunity for such checks for backwards
+        compatibility. It is called shortly after loading the opt file from
+        disk, and is called before the Agent is initialized.
+
+        Implementations of ``upgrade_opt`` should conform to high standards,
+        due to the risk of these methods becoming complicated. We recommend
+        the following behaviors:
+
+            1. ``upgrade_opt`` should only be used to provide backwards
+            compatibility.  Other behavior should find a different behavior.
+            2. Children should always call the parent's ``upgrade_opt`` first.
```

Comment:
any way we can enforce this?

Reply:
high standards in code review

Reply:
maybe tho? Pytorch has some way of knowing if you failed to call super().__init__()?

Reply:
https://github.com/pytorch/pytorch/blob/5b87049c66f0c64b405145ff57869de98729dae9/torch/nn/modules/module.py#L116-L118

here's how pytorch does it. it just looks for an initialized private variable, and then somewhere else that you don't override checks and complains.

Reply:
I actually kinda like that

Reply:
This turns out to be not possible here because it's a class agent, so any private variable would be a Singleton. :(